### PR TITLE
Refactor MergeBlocks to use iteration; adds Wasm GC support

### PR DESCRIPTION
**Pull request by kripken**
MergeBlocks basically does a generic operation to merge blocks
"to the outside",

```wat
(a
  (block
    (b)
    (c)
  )
)

=>

(block
  (b)
  (a
    (c)
  )
)
```
Often doing so lets the outer block be merged with an existing
one, and also it allows more optimizations on a's consumption of
c as now it is more direct. But MergeBlocks was written a very
long time ago, before the iteration API, so it had a bunch of
hardcoded things for specific instructions. In particular, that did
not handle GC.

This does a small refactoring to use iteration. The refactoring
is NFC, but while doing so it adds support for new relevant
instructions, including wasm GC.

This is already heavily tested, so almost no new tests are
added except for wasm GC. Also fuzzed over 100,000 iterations
just to be sure.

On j2cl output, this removes 20% of `ref.as*` operations, as it
turns out it's a common pattern to have a `ref.as_non_null` on
a block fallthrough; with this PR that block is moved out of the
way and then often we can optimize that away.

### CODE REVIEW

File: src/passes/MergeBlocks.cpp [hunk 1]
```diff
@@ -491,20 +494,26 @@ struct MergeBlocks : public WalkerPass<PostWalker<MergeBlocks>> {
     return outer;
   }
 
-  void visitUnary(Unary* curr) { optimize(curr, curr->value); }
-  void visitLocalSet(LocalSet* curr) { optimize(curr, curr->value); }
-  void visitLoad(Load* curr) { optimize(curr, curr->ptr); }
-  void visitReturn(Return* curr) { optimize(curr, curr->value); }
+  // Default optimizations for simple cases. Complex things are overridden
+  // below.
```
Comment (aheejin) [inline src/passes/MergeBlocks.cpp:498]: Are we able to override in `UnifiedExpressionVisitor`?

File: test/lit/passes/merge-blocks.wast [hunk 2]
```diff
@@ -31,4 +38,56 @@
...
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (struct.set $struct 0
+ ;; CHECK-NEXT:   (ref.null $struct)
+ ;; CHECK-NEXT:   (i32.const 5)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (nop)
+ ;; CHECK-NEXT: )
+ (func $struct.set
+  (block
+   (nop)
```
Comment (aheejin) [inline test/lit/passes/merge-blocks.wast:55]: Why are these necessary? And does this pass not only push `block`s outside but also remove unnecessary `block`s as well?

**Review by aheejin — APPROVED** [review]

File: src/passes/MergeBlocks.cpp (hunk 1 above)
Comment (kripken) [reply src/passes/MergeBlocks.cpp:498]: Yes. It defines default visitors `visitX() => visitExpression` for all `X`, but a subclass can override them.

File: test/lit/passes/merge-blocks.wast (hunk 2 above)
Comment (kripken) [reply test/lit/passes/merge-blocks.wast:55]: The nop is just to show more content in the outer block. Not necessary I guess, but I think it's a little clearer, as they remain afterwards, so we can see things in relation to them. But I can remove if you prefer?

Yes, the pass will merge blocks when it can. In fact the pushing of blocks outside is primarily in order to enable that merging, which is often possible.

**Review by tlively — APPROVED** [review]
Nice change! When reading through some of the other code, I saw this:

```
        // we are going to replace the block with the final element, so they
        // should be identically typed
        if (block->type != back->type) {
          return outer;
        }
```

That should probably be changed to a subtype check as well, possibly as a follow-up with another test.

**Comment by MaxGraey** [comment]
I found some minor regressions in optimization opportunities compare with prev implementation. Is it because current version using `isControlFlowStructure` now which pass only `Block`, `If`, `Loop`, `Try` but not `Switch`, `Drop`, `Break` ?
